module/mempool/stdmap/incorporated_result_seals_test.go

```diff
@@ -3,12 +3,191 @@ package stdmap
 import (
 	"testing"
 
+	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 
+	"pgregory.net/rapid"
+
 	"github.com/onflow/flow-go/model/flow"
 	"github.com/onflow/flow-go/utils/unittest"
 )
 
+// icrSealsMachine is a description of a state machine for testing IncorporatedresultSeals
+type icrSealsMachine struct {
+	icrs  *IncorporatedResultSeals       // icrSeals being tested
+	state []*flow.IncorporatedResultSeal // model of the icrSeals
+	limit uint
+}
+
+// Init is an action for initializing a icrSeals instance.
+func (m *icrSealsMachine) Init(t *rapid.T) {
+	n := uint(rapid.IntRange(1, 1000).Draw(t, "n").(int))
+	m.icrs = NewIncorporatedResultSeals(n)
+	m.limit = n
+}
+
+// Add is a conditional action which adds an item to the icrSeals.
+func (m *icrSealsMachine) Add(t *rapid.T) {
+	i := rapid.Int().Draw(t, "i").(int)
+
+	seal := unittest.IncorporatedResultSeal.Fixture(func(s *flow.IncorporatedResultSeal) {
+		s.Header.Height = uint64(i)
+	})
+
+	_, err := m.icrs.Add(seal)
+	require.NoError(t, err)
+
+	unmet := true
+	for _, v := range m.state {
+		if v.ID() == seal.ID() {
+			unmet = false
+		}
+	}
+
+	if unmet && seal.Header.Height >= m.icrs.lowestHeight {
+		m.state = append(m.state, seal)
+		if len(m.state) > int(m.limit) {
+			// we remove one of the max height elements if we're beyond the mempool limit
+			max_height := m.icrs.lowestHeight
+			for _, v := range m.state {
+				if v.Header.Height > max_height {
+					max_height = v.Header.Height
+				}
+			}
+			filtered_state := make([]*flow.IncorporatedResultSeal, 0)
+			for _, v := range m.state {
+				if v.Header.Height < max_height {
+					filtered_state = append(filtered_state, v)
+				} else {
```

Comment:
I think this logic is to implement the same ejection logic as the sealing mempool. 

The real ejection policy only remove exactly _one_ seal even if there are multiple seals at the same height, and the eject logic implemented here will remove _all_ seals at the same height.

maybe we could break it into two for loops:
1. the first for-loop will filter out exact one seal from `state`
2. the second for-loop will loop through the filtered_state and ensure each seal can be found.

Reply:
It seems we are testing the ejection behavior by actually implementing it. If so, is this how we usually do for property testing? I feel we are maintaining two copy of similar logic.

Reply:
The present logic does not remove all seals of maximal height (otherwise the test would not pass). It does check that the ejection removes one seal among those at maximal height.

This does maintain two similar copies of the logic, the code and the test.

But the code's logic is complex and spread in two places, the ejector (which select ejection candidates), and the actual ejection execution, defined as part of `BackData` and initialized in `Backend`.

The test's logic is simplified (we don't test for any particular block selection policy), self-contained, and expresses the core part of the property (the ejected block, whatever it is, is defined at maximal height).

Reply:
> And the second for-loop will keep all seals below the max height 4, in other words, remove all seals with height 4, then, the remaining seals are 2 and 3, which is different from the actual. Right?

The second for loop ends at line 69. In that second for loop, in the `else` case, we add to the model (`m.state`, here built through `filtered_state`) those seals of maximal height that are also kept by the mempool. The add is performed at line 66.

Reply:
Oh, I see. Nice! Thanks for explaining :)